File: src/main/scala/org/tindalos/principle/infrastructure/plugin/ConstraintsReader.scala
package org.tindalos.principle.infrastructure.plugin

import java.io.File

import org.apache.commons.io.FileUtils
import org.tindalos.principle.domain.constraints._
import org.tindalos.principle.domain.constraints.exception.InvalidConfigurationException
import org.tindalos.principle.domain.core.AnalysisPlan
import org.yaml.snakeyaml.Yaml

import scala.collection.JavaConverters._


object ConstraintsReader {

  //under src/main/resources
  private val defaultFileLocation = "/principle.yml"

  def readFromFile(fileLocation: Option[String]): AnalysisPlan = {
    val location = fileLocation.getOrElse(defaultFileLocation)
    fromYaml(readYAML(location), location)
  }

  private def fromYaml(yamlText: String, fileLocation: String): AnalysisPlan = {
    val yamlObject =
      new Yaml().load(yamlText).asInstanceOf[java.util.Map[String, Object]].asScala.toMap

    val rootPackage = yamlObject("root_package").asInstanceOf[String]

    val checksYaml = getYamlStructure(yamlObject, "checks").get

    val checks = {

      val modules: Option[SubmodulesBlueprint] = getYamlStructure(checksYaml, "modules")
          .map { modules ⇒
            val threshold = modules.get("violation_threshold").map(_.asInstanceOf[Int]).getOrElse(0)
            new SubmodulesBlueprint(fileLocation, threshold)
          }

      val packageCoupling = {
        val x: Option[(Option[RACD], Option[ADP])] = getYamlStructure(checksYaml, "package_coupling").map { pc ⇒

          val racdTh = pc.get("acd_threshold").map { threshold ⇒
            new RACD(threshold.asInstanceOf[Double])
          }
          val adpTh = pc.get("cyclic_dependencies_threshold").map { threshold ⇒
            new ADP(threshold.asInstanceOf[Int])
          }

          (racdTh, adpTh)
        }

        val grouping = yamlObject.get("structure_analysis_enabled")
            .filter(_.asInstanceOf[Boolean])
            .map { _ ⇒ Grouping.of()}

        val builder = PackageCouplingConstraints.builder()
        x.flatMap(_._2).foreach(adp => builder.adp(adp))
        x.flatMap(_._1).foreach(racd => builder.racd(racd))
        grouping.foreach(g => builder.grouping(g))
        builder.build()
      }

      new Constraints(
        getYamlStructure(checksYaml, "layering").map(toLayering).orNull,
        getYamlStructure(checksYaml, "third_party_restrictions").map(toThirdParty).map(tp => java.util.Optional.of(tp)).getOrElse(java.util.Optional.empty()),
        java.util.Optional.of(packageCoupling),
        modules.map(sm => java.util.Optional.of(sm)).getOrElse(java.util.Optional.empty()))
    }

    new AnalysisPlan(checks, rootPackage)
  }

  private def toThirdParty(structure: Map[String, Object]): ThirdParty = {
    val barriersYaml: Seq[Map[String, Object]] = structure("allowed_libraries")
        .asInstanceOf[java.util.List[java.util.Map[String, Object]]]
        .asScala.to[Seq].map(javaMap ⇒ javaMap.asScala.toMap)

    def toBarrier(m: Map[String, Object]): Barrier =
      new Barrier(
        m("layer").asInstanceOf[String],
        m("libraries").asInstanceOf[java.util.List[String]])

    new ThirdParty(
      barriersYaml.map(toBarrier).asJava,
      structure("violation_threshold").asInstanceOf[Int])
  }

  private def toLayering(structure: Map[String, Object]): Layering =
    new Layering(
      getYamList(structure, "layers").map(_.asJava).getOrElse(java.util.List.of()),
      structure.get("violation_threshold").map(_.asInstanceOf[Int]).getOrElse(0))

  private def getYamList(structure: Map[String, Object], field: String): Option[List[String]] =
    structure.get(field)
        .map(_.asInstanceOf[java.util.List[String]].asScala.to[List])

  private def getYamlStructure(structure: Map[String, Object], field: String): Option[Map[String, Object]] =
    structure.get(field)
        .map(_.asInstanceOf[java.util.Map[String, Object]].asScala.toMap)

  private def readYAML(fileLocation: String): String =
    try {
      FileUtils.readFileToString(new File(fileLocation))
    } catch {
      case ex: Exception =>
        ex.printStackTrace()
        throw new InvalidConfigurationException("problem with reading file from " + fileLocation);
    }

}

File: src/main/java/org/tindalos/guardrails/internal/infrastructure/core/ConstraintsReader.java
package org.tindalos.guardrails.internal.infrastructure.core;

import org.apache.commons.io.FileUtils;
import org.tindalos.guardrails.internal.domain.constraints.submodules.SubmoduleDefinitions;
import org.tindalos.guardrails.internal.domain.constraints.*;
import org.tindalos.guardrails.internal.domain.constraints.exception.InvalidConfigurationException;
import org.tindalos.guardrails.internal.domain.plan.AnalysisPlan;
import org.tindalos.guardrails.internal.infrastructure.analyzers.submodulesblueprint.YAMLBasedSubmodulesBlueprintProvider;
import org.yaml.snakeyaml.Yaml;

import java.io.File;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Reads and parses the Principle configuration YAML file into an {@link AnalysisPlan}.
 * Supports layering, third-party restrictions, package coupling, and submodule definitions.
 */
public class ConstraintsReader {

    private static final String DEFAULT_FILE_LOCATION = "/guardrails.yml";

    public static AnalysisPlan readFromFile(Optional<String> fileLocation) {
        var location = fileLocation.orElse(DEFAULT_FILE_LOCATION);
        return fromYaml(readYAML(location), location);
    }

    @SuppressWarnings("unchecked")
    private static AnalysisPlan fromYaml(String yamlText, String fileLocation) {
        Map<String, Object> yamlObject = (Map<String, Object>) new Yaml().load(yamlText);

        var rootPackage = (String) yamlObject.get("root_package");

        var constraintsYaml = getYamlStructure(yamlObject, "constraints").orElseThrow();

        var modules = parseModules(constraintsYaml, rootPackage, fileLocation);
        var packageCoupling = parsePackageCoupling(constraintsYaml, yamlObject);

        var constraints = new Constraints(
                getYamlStructure(constraintsYaml, "layering").map(ConstraintsReader::toLayering),
                getYamlStructure(constraintsYaml, "third_party_restrictions").map(ConstraintsReader::toThirdParty),
                Optional.of(packageCoupling),
                modules);

        return new AnalysisPlan(constraints, rootPackage);
    }

    @SuppressWarnings("unchecked")
    private static Optional<SubmoduleDefinitions> parseModules(Map<String, Object> constraintsYaml,
                                                                String rootPackage,
                                                                String fileLocation) {
        return getYamlStructure(constraintsYaml, "modules")
                .filter(m -> m.containsKey("module-definitions"))
                .map(modules -> {
                    var threshold = Optional.ofNullable((Integer) modules.get("violation_threshold")).orElse(0);
                    return new YAMLBasedSubmodulesBlueprintProvider()
                            .readSubmoduleDefinitions(rootPackage, fileLocation, threshold);
                });
    }

    @SuppressWarnings("unchecked")
    private static PackageCouplingConstraints parsePackageCoupling(Map<String, Object> constraintsYaml,
                                                                    Map<String, Object> yamlObject) {
        var builder = PackageCouplingConstraints.builder();

        getYamlStructure(constraintsYaml, "package_coupling").ifPresent(pc -> {
            Optional.ofNullable(pc.get("acd_threshold"))
                    .map(t -> new RACD((Double) t))
                    .ifPresent(builder::racd);
            Optional.ofNullable(pc.get("cyclic_dependencies_threshold"))
                    .map(t -> new ADP((Integer) t))
                    .ifPresent(builder::adp);
        });

        Optional.ofNullable(yamlObject.get("structure_analysis_enabled"))
                .filter(v -> (Boolean) v)
                .map(v -> Grouping.of())
                .ifPresent(builder::grouping);

        return builder.build();
    }

    @SuppressWarnings("unchecked")
    private static ThirdParty toThirdParty(Map<String, Object> structure) {
        var barriersYaml = (List<Map<String, Object>>) structure.get("allowed_libraries");
        var barriers = barriersYaml.stream()
                .map(m -> new Barrier(
                        (String) m.get("layer"),
                        (List<String>) m.get("libraries")))
                .toList();
        return new ThirdParty(barriers, (Integer) structure.get("violation_threshold"));
    }

    @SuppressWarnings("unchecked")
    private static Layering toLayering(Map<String, Object> structure) {
        var layers = Optional.ofNullable((List<String>) structure.get("layers"))
                .orElse(List.of());
        var threshold = Optional.ofNullable((Integer) structure.get("violation_threshold")).orElse(0);
        return new Layering(layers, threshold);
    }

    @SuppressWarnings("unchecked")
    private static Optional<Map<String, Object>> getYamlStructure(Map<String, Object> structure, String field) {
        return Optional.ofNullable((Map<String, Object>) structure.get(field));
    }

    private static String readYAML(String fileLocation) {
        try {
            return FileUtils.readFileToString(new File(fileLocation), java.nio.charset.StandardCharsets.UTF_8.toString());
        } catch (Exception ex) {
            ex.printStackTrace();
            throw new InvalidConfigurationException("problem with reading file from " + fileLocation);
        }
    }
}

